webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/services/impl/GitTechniqueReader.scala

```diff
@@ -181,51 +173,55 @@ class GitTechniqueReader(
-    }
+  private[this] val currentTechniquesInfoCache: Ref[TechniquesInfo] = semaphore.withPermit(
+    for {
+      currentRevTree <- revisionProvider.currentRevTreeId
+      db             <- repo.db
+      res            <- processRevTreeId(db, currentRevTree).catchAll {
+                          case err @ SystemError(m, NoRootCategory(msg)) =>
+                            for {
+                              _            <- TechniqueReaderLoggerPure.error(s"The stored Git revision '${currentRevTree}' does not provide a root category.xml, which is mandatory. Error message was: ${err.fullMsg}")
+                              newRevTreeId <- revisionProvider.getAvailableRevTreeId
+                              res <- if(newRevTreeId != currentRevTree) {
+                                        TechniqueReaderLoggerPure.error(s"Trying to load last available revision of the technique library to unstuck the situation.") *>
+                                        revisionProvider.setCurrentRevTreeId(newRevTreeId) *>
+                                        processRevTreeId(db, newRevTreeId)
+                                      } else {
+                                        Inconsistancy("Please add a root category.xml and commit it before restarting Rudder.").fail
+                                      }
+                            } yield res
+                          case e:MissingObjectException => //ah, that commit is not know on our repos
+                            TechniqueReaderLoggerPure.error("The stored Git revision for the last version of the known Technique Library was not found in the local Git repository. " +
+                                "That may happen if a commit was reverted, the Git repository was deleted and created again, or if LDAP datas where corrupted. Loading the last available Techique library version.") *>
+                            revisionProvider.getAvailableRevTreeId.flatMap(newRevTreeId =>
+                              revisionProvider.setCurrentRevTreeId(newRevTreeId) *>
+                              processRevTreeId(db, newRevTreeId)
+                            )
+                        }
+    } yield res
+  ).flatMap(Ref.make(_)).runNow
+
+  private[this] val nextTechniquesInfoCache: Ref[(ObjectId,TechniquesInfo)] = {
+    (for {
+      a <- revisionProvider.currentRevTreeId
+      b <- currentTechniquesInfoCache.get
+      r <- Ref.make((a, b))
+    } yield r).runNow
   }
 
-  private[this] var nextTechniquesInfoCache : (ObjectId,TechniquesInfo) = (revisionProvider.currentRevTreeId.runNow, currentTechniquesInfoCache)
   //a non empty list IS the indicator of differences between current and next
-  private[this] var modifiedTechniquesCache : Map[TechniqueName, TechniquesLibraryUpdateType] = Map()
+  private[this] val modifiedTechniquesCache: Ref[Map[TechniqueName, TechniquesLibraryUpdateType]] = Ref.make(Map[TechniqueName, TechniquesLibraryUpdateType]()).runNow
```

Comment:
In place of `var` we now use `Ref`, which are atomically accessed memory zone (with `get`, `set` or `update`)